app/main.py

```diff
@@ -1,214 +1,75 @@
 
-    # apply weapon
-    mordred["power"] += mordred["weapon"]["power"]
-
-    # apply potion if exist
-    if mordred["potion"] is not None:
-        if "power" in mordred["potion"]["effect"]:
-            mordred["power"] += mordred["potion"]["effect"]["power"]
-
-        if "protection" in mordred["potion"]["effect"]:
-            mordred["protection"] += mordred["potion"]["effect"]["protection"]
-
-        if "hp" in mordred["potion"]["effect"]:
-            mordred["hp"] += mordred["potion"]["effect"]["hp"]
-
-    # red_knight
-    red_knight = knightsConfig["red_knight"]
-
-    # apply armour
-    red_knight["protection"] = 0
-    for a in red_knight["armour"]:
-        red_knight["protection"] += a["protection"]
+    return {
+        knight.name: knight.hp
+        for knight in knights
+    }
 
-    # apply weapon
-    red_knight["power"] += red_knight["weapon"]["power"]
 
-    # apply potion if exist
-    if red_knight["potion"] is not None:
-        if "power" in red_knight["potion"]["effect"]:
-            red_knight["power"] += red_knight["potion"]["effect"]["power"]
+def _create_knights(knights_config: dict) -> list:
+    knights = []
 
-        if "protection" in red_knight["potion"]["effect"]:
-            red_knight["protection"] += red_knight["potion"]["effect"]["protection"]
+    for stats in knights_config.values():
+        knight = Knight(
+            stats["name"],
+            stats["power"],
+            stats["hp"],
+            stats["armour"],
+            stats["weapon"],
+            stats["potion"]
+        )
+        knights.append(knight)
 
-        if "hp" in red_knight["potion"]["effect"]:
-            red_knight["hp"] += red_knight["potion"]["effect"]["hp"]
+    return knights
 
-    # -------------------------------------------------------------------------------
-    # BATTLE:
 
-    # 1 Lancelot vs Mordred:
-    lancelot["hp"] -= mordred["power"] - lancelot["protection"]
-    mordred["hp"] -= lancelot["power"] - mordred["protection"]
+def _duel_battle(knight_1: Knight, knight_2: Knight) -> None:
+    knight_1.hp = knight_1.hp - (knight_2.power - knight_1.protection)
+    _check_hp([knight_1, knight_2])
 
-    # check if someone fell in battle
-    if lancelot["hp"] <= 0:
-        lancelot["hp"] = 0
 
-    if mordred["hp"] <= 0:
-        mordred["hp"] = 0
+def _check_hp(knights: list) -> None:
+    for knight in knights:
+        if knight.hp <= 0:
+            knight.hp = 0
 
-    # 2 Arthur vs Red Knight:
-    arthur["hp"] -= red_knight["power"] - arthur["protection"]
-    red_knight["hp"] -= arthur["power"] - red_knight["protection"]
 
-    # check if someone fell in battle
-    if arthur["hp"] <= 0:
-        arthur["hp"] = 0
+def _apply_armour(knights: list) -> None:
+    for knight in knights:
+        protection = 0
+        if knight.armour:
+            for armour in knight.armour:
+                protection += armour.protection
+            knight.protection = protection
```

Comment:
let's simplify it. `if not ...: next iteration`. In this way loop will be moved out of if statement. it's called cognitive complexity when you have many nested ifs and loops. it's better to get rid of such constructions